File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
       int n = nums.length;

        // Step 1: Initialize the answer with the smallest possible value
        int ans = Integer.MIN_VALUE;

        // Step 2: Initialize prefix and suffix products
        int prefix = 1;
        int suffix = 1;

        // Step 3: Traverse from both directions
        for (int i = 0; i < n; i++) {

            // Step 4: If prefix or suffix becomes 0, reset it to 1
            // Because product with 0 restarts the subarray
            if (prefix == 0) prefix = 1;
            if (suffix == 0) suffix = 1;

            // Step 5: Update prefix (left-to-right product)
            prefix *= nums[i];

            // Step 6: Update suffix (right-to-left product)
            suffix *= nums[n - i - 1];

            // Step 7: Update the answer with the max of current products
            ans = Math.max(ans, Math.max(prefix, suffix));
        }

        // Step 8: Return the maximum product found
        return ans;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    let ans = Number.MIN_SAFE_INTEGER;

    let n = nums.length;

    let prefix = 1;
    let suffix = 1;
    for(let i = 0; i < n; i++) {
        if(prefix == 0) {
            prefix = 1;
        }
        if(suffix == 0) {
            suffix = 1;
        }

        prefix = prefix * nums[i];
        suffix = suffix * nums[n - i -1];

        ans = Math.max(ans, prefix, suffix);
    }

    if(ans == -0) {
        return 0;
    }
    return ans;
};